llvm/docs/DirectX/DXILOpTableGenDesign.rst

```diff
@@ -14,147 +14,254 @@ Introduction
-4. DXIL Operation Class is represented as follows
+2. A documentation string for the operation.
+3. The LLVM Intrinsic that maps to the operation is represented as ``Intrinsic`` defined in
+   `Intrinsics.td <https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/IR/Intrinsics.td>`_.
+4. The unique operation id is represented by an integer.
+5. DXIL Operation Class is represented as follows
 
    .. code-block::
 
         // Abstraction of DXIL Operation class.
-        // It encapsulates an associated function signature viz.,
-        // returnTy(param1Ty, param2Ty, ...) represented as a list of LLVMTypes.
-        // DXIL Ops that belong to a DXILOpClass record the signature of that DXILOpClass
-
-        class DXILOpClass<list<LLVMType> OpSig> {
-          list<LLVMType> OpSignature = OpSig;
-        }
+        class DXILOpClass;
 
-   Concrete operation classes, such as ``unary`` are defined inheriting from ``DXILOpClass``.
-5. Valid overload types are represented as a list of ``LLVMType``.
-6. Concrete records of DXIL versions and are defined by inheriting from the class
+   Concrete operation records, such as ``unary`` are defined by inheriting from ``DXILOpClass``.
+6. Return and argument types of the operation are represented as ``dag``s using the
+   special markers ``out`` and ``ins``. An overload type, if supported by the operation, is
+   denoted as the positional type ``dxil_overload_ty`` in the argument or in the result, where
+   ``dxil_overload_ty`` is defined to be synonymous to ``llvm_any_ty``.
 
    .. code-block::
 
-        // Abstract class to represent major and minor version values
-        class Version<int major, int minor> {
-          int Major = major;
-          int Minor = minor;
-        }
+      defvar dxil_overload_ty = llvm_any_ty
 
-7. A documentation string for the operation.
 
+7. Valid overload types and shader stages predicated on Shader Model version are specified
+   as a list of ``Constraint`` records. Representation of ``Constraints`` class is described
+   a later section.
+8. Various attributes of the DXIL Operation that are not predicated on Shader Model version
+   are represented as a ``dag`` using the special marker ``attrs``. Representation of ``Attributes`` 
+   class is described in a later section.
 
 A DXIL Operation is represented by the following TableGen class by encapsulating the various
-TableGen representations of its properties described above.
+TableGen representations of its attributes described above.
+
+.. code-block::
+
+   // Abstraction DXIL Operation
+   class DXILOp {
+     // A short description of the operation
+     string Doc = "";
+
+     // Opcode of DXIL Operation
+     int OpCode = 0;
+
+     // Class of DXIL Operation.
+     DXILOpClass OpClass = UnknownOpClass;
+
+     // LLVM Intrinsic DXIL Operation maps to
+     Intrinsic LLVMIntrinsic = ?;
+
+     // Dag containing the arguments of the op. Default to 0 arguments.
+     dag arguments = (ins);
```

Comment:
It might be nice to have an example that actually sets the `arguments` and `result` fields so that we can see what it looks like.

Reply:
> It might be nice to have an example that actually sets the `arguments` and `result` fields so that we can see what it looks like.

Added an example. Thanks!